Allow controllers to claim resources from multiple interfaces

adolfo-rt opened the pull request:
Fixes #75 and #208.

This PR contains the core changes required to allow controllers to claim resources from more than one hardware interface. It targets _jade-devel_ because it breaks both C++ and ROS APIs (see #201 ). The changes are very localized, and mostly affect:
- The `controller_manager/ListControllers` service.
- The `initRequest` method of `ControllerBase` (mostly power users).

The work isn't finalized yet, pending tasks:
- [x] Add more tests exercising the new features
- [x] Propagate the changes to `rqt_controller_manager`
- [x] Propagate the changes to `ros_controllers`

The timeline is to get this reviewed and (if the review is positive) merged for the official jade release.

What will not be in the scope of this work package is a `ControllerBase` subclass allowing for more than one HW interface, i.e., a multi-interface analog of `Controller`. Since that is an API addition, it can be iterated on with more time and incorporated later in the jade cycle.

Commit:
hardware_interface: Multi-interface controllers

- This is a C++ API breaking changeset.

- Modify ControllerInfo class to allow controllers to claim resources from
  multiple hardware interfaces.

- Propagate changes to RobotHW::checkForConflict: Default resource ownsership
  policy is aware of controllers claiming resources from  multiple hardware
  interfaces.

- Update and extend the corresponding test suite.

Commit:
controller_interface: Multi-interface controllers

- This is a C++ API breaking changeset.

- Modify controller interfaces to allow for controllers that claim resources
  from more than one hardware interface.

Commit:
controller_manager_msgs: Multi-interface controllers

- This is a ROS API and (internal) Python API breaking changeset.

- Modify and extend ROS message definitions to allow for controllers that
  claim resources from more than one hardware interface.

- Modify Python helpers and update the corresponding test suite to take into
  account above changes.

Commit:
controller_manager: Multi-interface controllers

- This is a C++ API breaking changeset.

- Make controller_manager aware of controllers that claim resources from more
  than one hardware interface.

- Update and extend the corresponding test suite.

Commit:
controller_interface: Remove pure virtual method

- This is a C++ API breaking changeset.

- Remove getHardwareInterfaceType() pure virtual method from ControllerBase
  class. This public method assumes a single hardware interface, and is only
  used by the Controller subclass.

- Controller class still preserves the method, albeit protected and non-virtual.

davetcoleman commented:
I read through the changes just now and, although I don't understand all of it, it looks like a quality PR without any obvious mistakes. Looks like you put effort into making things cleaner.

sachinchitta commented on controller_interface/include/controller_interface/controller_base.h:
```diff
@@ -122,12 +126,11 @@ class ControllerBase
 
-  /// Get the name of this controller's hardware interface type
-  virtual std::string getHardwareInterfaceType() const = 0;
+  typedef std::vector<hardware_interface::InterfaceResources> ClaimedResourcesType;
```
Why call this a "Type" - its a list of resources (and associated hardware interfaces). Maybe just call it ClaimedResources?

adolfo-rt replied:
Agreed

sachinchitta commented on controller_manager/src/controller_manager.cpp:
```diff
@@ -225,7 +225,7 @@ bool ControllerManager::loadController(const std::string& name)
   ROS_DEBUG("Initializing controller '%s'", name.c_str());
   bool initialized;
-  std::set<std::string> claimed_resources; // Gets populated during initRequest call
+  std::vector<hardware_interface::InterfaceResources> claimed_resources; // Gets populated during initRequest call
```
Use the typedef above for consistency?

adolfo-rt replied:
Agreed

sachinchitta commented on controller_manager/test/hwi_switch_test.cpp:
```diff
@@ -151,19 +152,24 @@ class SwitchBot : public hardware_interface::RobotHW
-            for (std::set<std::string>::const_iterator res_it = it->resources.begin(); res_it != it->resources.end(); ++res_it)
+            if (it->claimed_resources.size() != 1)
+            {
+                ROS_FATAL("We expect controllers to claim resoures from only one interface. This should never happen!");
```
This will go away when controllers starting claiming resource from more than one interface (in the future)?

adolfo-rt replied:
This test only exercises the joint mode switching behavior, which is orthogonal to controllers with resources from multiple interfaces. For simplicity, I think it's fine to preserve the assumption that this test works only with controllers whose resources come from a single interface.

I did the minimal changes to this test to make it compatible with this PR.

Still, to take your feedback into account, I've added a [comment](https://github.com/ros-controls/ros_control/pull/204/files#diff-ace4c98502cb1cc5affcd660cd076859R61) in the test stating its assumptions, to prevent potential confusion.

sachinchitta commented:
LGTM.

adolfo-rt commented:
@sachinchitta, thanks for the review. I've addressed your points, which show up as two changesets containing 'Sachin Chitta code review' in the summary. I'll squash them with their parent before merging.

Commit:
controller_interface: Add MultiInterfaceController

MultiInterfaceController: Subclass of ControllerBase which allows to claim
resources from up to four different hardware interfaces.

It's similar in spirit to the single-interface Controller class, in that
hardware interfaces are specified as template parameters. One difference is that
the custom controller initialization init() methods are passed a RobotHW*
parameter containing only the hardware interfaces requested by the controller
(as opposed to the actual hardware interface for Controller, or a RobotHW*
representing the _entire_ robot for ControllerBase).

Requested hardware interfaces are required by default, but can be made optional
through a constructor flag.

Commit:
controller_manager_tests: Extend test suite

- Exercise much more of the controller_manager ROS API.

- Create multi-interface test controllers and exercise them in tests.

Commit:
hardware_interface: Add InterfaceManager::getNames

Add new method that allows to query the names of all interfaces managed by
an InterfaceManager instance.

adolfo-rt commented:
> What will not be in the scope of this work package is a ControllerBase subclass allowing for more than one HW interface, i.e., a multi-interface analog of Controller. Since that is an API addition, it can be iterated on with more time and incorporated later in the jade cycle.

I went against this statement and implemented a proposal for a **multi-interface controller API**. I felt that a decent reference implementation would allow to make design shortcomings evident. So far so good :-). If you want to browse the proposed API, please look at the [MultiInterfaceController](https://github.com/ros-controls/ros_control/blob/dbb8e77a189796f7f5838692afd44c1d23025f34/controller_interface/include/controller_interface/multi_interface_controller.h) class. One nifty power-user feature is that **interfaces can be made optional**, so you can write a controller that for instance requires two interfaces, one of which is optional, and its absence won't prevent the controller from initializing and starting.

I also took the opportunity in 2b0a823 to **unit test** this new class, and to test the `controller_manager` more thoroughly, in general.

Commit:
transmission_interface: Add new loading method

Allow loading transmissions from a vector of TransmissionInfo instances.

adolfo-rt commented:
Update: I've already propagated these changes to `ros_controllers`, only the `rqt_joint_trajectory_controller` was affected. I'm now tackling `rqt_controller_manager`, which is the last known blocker to merge this PR and allow the official jade release.

Commit:
Add helper to query rosparam controller configs

There is no way to identify uninitialized controllers other than by inspecting
the ROS parameter server, and looking for the required controller config
structure, which is the existence of the <ctrl_name>/type parameter.

This changeset adds Python helpers for performing this query.

Commit:
rqt_cm: Multi-interface controllers, UI revamp

- Make the rqt_controller_manager aware of multi-interface controllers.

- Reduce the amount of screen real-estate used by the plugin. Controller
  information that was previously shown in the plugin widget has been
  deferred to a pop-up that appears when double-clicking a controller.

- Make main view read-only. To affect the state of available controllers
  (load, unload,start, stop), one must take explicit action: right-click
  context menu. This was already so to a certain exxtent, but you could
  still load and start uninitialized controllers from the main view.

- Show uninitialized controllers (fetched from parameter server) in the same
  list as stopped and running controllers.

- Make less assumptions when finding running controller managers. Use
  existing helpers on controller_mamager_msgs.utils Python module.

Commit:
rqt_cm: Allow running as standalone application

adolfo-rt commented:
I consider this PR feature-complete. Standby for a Jade release in the coming days. I want the dust to settle a bit and merge outstanding PRs in indigo, and forward-port them to jade.

adolfo-rt commented:
Please take a moment to review #208, which is the last bit that was remaining.

Commit:
Merge pull request #206 from ros-controls/load-trans

transmission_interface: Add new loading method

adolfo-rt commented:
Giving this a go (!)
